promql/engine.go

```diff
@@ -861,6 +866,8 @@ func (ev *evaluator) eval(expr Expr) Value {
 				if len(outVec) > 0 {
 					ss.Points = append(ss.Points, Point{V: outVec[0].Point.V, T: ts})
 				}
+				// Only buffer stepRange milliseconds from the second step on.
```

Comment:
This doesn't sound right at all. We need all of the range.

Reply:
Assuming all data from the previous step is still in `points`, then strictly speaking we don't need any of the data before the previous step to also be buffered by `it`. We can "merge" the samples in `points` with the following `stepRange` worth of samples and end up with the same data that `it` with a delta of `selRange` would otherwise provide.

As stated in the PR description, it's iterating over all the overlapping data that slows things down, not so much copying it into `points`. So the intent is to essentially reduce the data buffered by `BufferedSeriesIterator` to the bare minimum (i.e. the minimum between `selRange` and `ev.interval`) from the second step on and complementing that data with what's already in the slice.

Ideally one would pass the iterator's buffer straight through to the eval logic (no copying involved as it's not, strictly speaking, needed) but that's not easily done without major refactoring.

Reply:
That looks cleaner now. Still a tad convoluted to have the reducedelta call interacting with the matrixIterSlice.

Reply:
Hmm... Now that you bring it up, I could have `matrixIterSlice` adjust the delta directly instead, but I'm not actually sure that would make it cleaner. Or safer.

Reply:
Yeah, as you say a deeper change so we don't have to do all this copying would be better but a bit much for now.